### PR TITLE
Refactor display_login_attempts rule for simplicity and avoid noise

### LINKED ISSUE #9031: Improve the description of display_login_attempts rule and respective remediation
#### Description of problem:

The `display_login_attempts` is mentioning the following statement in its description:
```
And make sure that the silent option is not set for pam_lastlog module.
```
- https://github.com/ComplianceAsCode/content/blob/master/linux_os/guide/system/accounts/accounts-pam/display_login_attempts/rule.yml#L17

This is probably inherited from STIG Requirement:
- https://www.stigviewer.com/stig/red_hat_enterprise_linux_8/2021-12-03/finding/V-230381
- https://www.stigviewer.com/stig/red_hat_enterprise_linux_8/2021-12-03/finding/V-230382

As it is described the most probable interpretation is that `silent` option should be removed in all `pam_lastlog` lines.

However, the intention of the requirement is to make the user aware of last access using its account and consequently giving him more information to notice any suspected access.

This could be achieved without removing the `silent` option from all lines. There are cases where the `silent` option is desired and safe to be kept while honoring the requirement goal. Removing it in all lines impact in user experience by creating avoidable noise.

#### SCAP Security Guide Version:

master

#### Operating System Version:

N/A

#### Steps to Reproduce:

1. Apply the `display_login_attempts` remediation
2. Login in the system and not that the last successful login information is duplicated.

#### Actual Results:

Last successful login information is duplicated.

#### Expected Results:

Last successful login information is not duplicated.

#### Additional Information/Debugging Steps:

This is the usual content of `/etc/pam.d/postlogin` file before the remediation:
```
session     optional                   pam_umask.so silent
session     [success=1 default=ignore] pam_succeed_if.so service !~ gdm* service !~ su* quiet
session     [default=1]                pam_lastlog.so nowtmp showfailed
session     optional                   pam_lastlog.so silent noupdate showfailed
```
In this default configuration, this is the output after a failed login followed by a successful login:
```
$ ssh root@rhel8
Last failed login: Wed Aug  9 08:31:44 EDT 2023 from 192.168.122.1 on ssh:notty
There was 1 failed login attempt since the last successful login.
Last login: Wed Aug  9 08:31:00 2023 from 192.168.122.1
```

This would be the content of `/etc/pam.d/postlogin` file after the remediation:
```
session     required                  pam_lastlog.so showfailed
session     optional                   pam_umask.so silent
session     [success=1 default=ignore] pam_succeed_if.so service !~ gdm* service !~ su* quiet
session     [default=1]                pam_lastlog.so nowtmp showfailed
session     optional                   pam_lastlog.so noupdate showfailed
```
It means that failed logins will be shown in any case, but successful login will be shown three times, one for each `pam_lastlog.so` line. The inclusion of the first line is even unnecessary to achieve requirement. Also, removing the silent in this last line will only create noise by repeating again information to the user, as shown in the example:

```
$ ssh root@rhel8
Last login: Wed Aug  9 08:31:52 EDT 2023 from 192.168.122.1 on pts/0
Last failed login: Wed Aug  9 08:35:39 EDT 2023 from 192.168.122.1 on ssh:notty
There were 3 failed login attempts since the last successful login.
Last login: Wed Aug  9 08:35:44 EDT 2023 from 192.168.122.1 on ssh
Last login: Wed Aug  9 08:35:44 2023 from 192.168.122.1
```

By not removing the `silent` option on the last line, this would be the content of `/etc/pam.d/postlogin` file after the remediation:
```
session     required                  pam_lastlog.so showfailed
session     optional                   pam_umask.so silent
session     [success=1 default=ignore] pam_succeed_if.so service !~ gdm* service !~ su* quiet
session     [default=1]                pam_lastlog.so nowtmp showfailed
session     optional                   pam_lastlog.so silent noupdate showfailed
```
It means that failed logins will be shown in any case, but successful login won't be shown three times, as shown in the example:

```
$ ssh root@rhel8
Last login: Wed Aug  9 08:35:44 EDT 2023 from 192.168.122.1 on pts/0
Last failed login: Wed Aug  9 08:39:51 EDT 2023 from 192.168.122.1 on ssh:notty
There were 3 failed login attempts since the last successful login.
Last login: Wed Aug  9 08:39:55 2023 from 192.168.122.1
```

This is still duplicating information without any additional value. After some tests it was concluded that default settings are pretty much reasonable and no changes are necessary.

#### Addition Information/Debugging Steps:
- https://bugzilla.redhat.com/show_bug.cgi?id=2157044
- https://github.com/ComplianceAsCode/content/pull/9009
- https://www.stigviewer.com/stig/red_hat_enterprise_linux_8/2022-12-06/finding/V-230381
- https://linux.die.net/man/8/pam_lastlog

### PR DESCRIPTION
**Pull request by marcusburghardt**
#### Description:

The `pam_lastlog.so` module is used in the `/etc/pam.d/postlogin` file (or login file depending on the distro) to display `Last login` information for certain programs. Although this module is only explicitly mentioned in the `postlogin` file, it is indirectly used by various other PAM configuration files, such as those for `login`, `ssh`, `passwd`, and others. The reason for this is to ensure that the `Last login` information is shown for programs like `gdm` and `su`, while avoiding redundancy for programs that already display this information without needing `pam_lastlog.so`. Despite the usage of `silent` option in one line, the default configuration of `/etc/pam.d/postlogin` ensures consistent display of `Last login` information for specific programs.

In short we only need to check if at least one line contains `am_lastlog.so`, `showfailed` and not `silent`.

#### Rationale:

- Simplify the check
- Do not remediate what is already working as expected
- Avoid noise and confusion to users by not duplicating information
- Fixes #9031
- Fixes https://bugzilla.redhat.com/show_bug.cgi?id=2157044

#### Review Hints:

The commits were organized in a sequence to facilitate the review.
The most import is actually to understand the reasons behind the changes, which are better contextualized in #9031
Finally, the automatus should work fine with to validate the technical aspects of the check and remediation

### CODE REVIEW

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/ansible/shared.yml [hunk 1]
```diff
@@ -4,10 +4,11 @@
...
-{{{ ansible_remove_pam_module_option_configuration('/etc/pam.d/login', 'session', '', 'pam_lastlog.so', 'silent') }}}
+{{% if product in ["sle12", "sle15"] or 'ubuntu' in product %}}
+{{% set pam_lastlog_path = "/etc/pam.d/login" %}}
 {{% else %}}
-{{{ ansible_ensure_pam_module_configuration('/etc/pam.d/postlogin', 'session', 'required', 'pam_lastlog.so', 'showfailed', '', 'BOF') }}}
-{{{ ansible_remove_pam_module_option_configuration('/etc/pam.d/postlogin', 'session', '', 'pam_lastlog.so', 'silent') }}}
+{{% set pam_lastlog_path = "/etc/pam.d/postlogin" %}}
 {{% endif %}}
+
+{{{ ansible_ensure_pam_module_configuration(pam_lastlog_path, 'session', '[default=1]', 'pam_lastlog.so', 'showfailed', '', 'BOF') }}}
```
Comment (teacup-on-rockingchair) [inline linux_os/guide/system/accounts/accounts-pam/display_login_attempts/ansible/shared.yml:13]: This seem to fail to insert the showfailed line in case I have a commented out line in my config like:
```
#%PAM-1.0
auth      requisite  pam_nologin.so
auth      include    common-auth
account   include    common-account
password  include    common-password
session   required   pam_loginuid.so
session   optional   pam_keyinit.so force revoke
session   include    common-session
#session  optional   pam_lastlog.so nowtmp showfailed
session   optional   pam_mail.so standard
auth    required    pam_tally2.so deny=3 onerr=fail
```

**Review by teacup-on-rockingchair — CHANGES_REQUESTED** [review]
Beside the ansible issue saw no other definately looks better :+1:

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/ansible/shared.yml (hunk 1 above)
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/ansible/shared.yml:13]: @teacup-on-rockingchair , I tried to apply the remediation in a file with this content and the line was inserted at the beginning of the file (BOF):
```
session    [default=1]    pam_lastlog.so
#%PAM-1.0
auth      requisite  pam_nologin.so
auth      include    common-auth
account   include    common-account
password  include    common-password
session   required   pam_loginuid.so
session   optional   pam_keyinit.so force revoke
session   include    common-session
#session  optional   pam_lastlog.so nowtmp showfailed
session   optional   pam_mail.so standard
auth    required    pam_tally2.so deny=3 onerr=fail
```
I see the line used for remediation and even the `after_match` regex passed to the macro might be different among distros. I can create a jinja2 condition in these remediation and make it more flexible to be maintained.

Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/ansible/shared.yml:13]: Thanks @teacup-on-rockingchair . Investigating this case I noticed the remediation can be improved a little more.

**Review by teacup-on-rockingchair — APPROVED** [review]
Nice stuff :bow:

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/ansible/shared.yml (hunk 1 above)
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/ansible/shared.yml]: It should be fine now @teacup-on-rockingchair

**Review by Mab879 — APPROVED** [review]
LGTM. Automatus tests locally in RHEL7-RHEL9 VMs.

**Comment by marcusburghardt** [comment]
@dodys and @freddieRv could you also take a look on this PR, please?

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/rule.yml [hunk 2]
```diff
@@ -57,20 +54,20 @@ references:
...
-ocil_clause: '"pam_lastlog" is missing from "{{{ pam_lastlog_path }}}" file, or the silent option is present'
+ocil_clause: '"pam_lastlog.so" is not properly configured in "{{{ pam_lastlog_path }}}" file'
 
 ocil: |-
     Verify users are provided with feedback on when account accesses last occurred with the following command:
 
     <pre>$ sudo grep pam_lastlog {{{ pam_lastlog_path }}}
 
-    session required pam_lastlog.so showfailed</pre>
+    session [default=1] pam_lastlog.so nowtmp showfailed</pre>
```
Comment (dodys) [inline linux_os/guide/system/accounts/accounts-pam/display_login_attempts/rule.yml:64]: for now the Ubuntu 20.04 STIG benchmark still mentions:
`session required pam_lastlog.so showfailed`

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/rule.yml [hunk 3]
```diff
@@ -57,20 +54,20 @@ references:
...
 
 fixtext: |-
-    Configure {{{ full_name }}} to provide users with feedback on when account accesses last occurred by setting the required configuration options in "/etc/pam.d/postlogin".
+    Configure {{{ full_name }}} to provide users with feedback on when account accesses last occurred by setting the required configuration options in "{{{ pam_lastlog_path }}}".
 
-    Add the following line to the top of "/etc/pam.d/postlogin":
+    Add the following line to the top of "{{{ pam_lastlog_path }}}":
 
-    session required pam_lastlog.so showfailed
+    session [default=1] pam_lastlog.so nowtmp showfailed
```
Comment (dodys) [inline linux_os/guide/system/accounts/accounts-pam/display_login_attempts/rule.yml:71]: for now the Ubuntu 20.04 STIG benchmark still mentions:
`session required pam_lastlog.so showfailed`

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/correct_value.pass.sh [hunk 4]
```diff
@@ -1,10 +1,11 @@
 #!/bin/bash
-# platform = Red Hat Enterprise Linux 7,multi_platform_ol,multi_platform_ubuntu
+# platform = multi_platform_ol,multi_platform_ubuntu,Red Hat Enterprise Linux 7
```
Comment (dodys) [inline linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/correct_value.pass.sh]: missing SLE platform?

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/default_config.pass.sh [hunk 5]
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
-# platform = Oracle Linux 7,Red Hat Enterprise Linux 7,multi_platform_ubuntu
+# platform = multi_platform_ubuntu,Oracle Linux 7,Red Hat Enterprise Linux 7
```
Comment (dodys) [inline linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/default_config.pass.sh]: missing SLE platform?

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/ansible/shared.yml [hunk 6]
```diff
@@ -4,10 +4,14 @@
...
-{{% if product in ["sle12", "sle15"] or "ubuntu" in product %}}
-{{{ ansible_ensure_pam_module_configuration('/etc/pam.d/login', 'session', 'required', 'pam_lastlog.so', 'showfailed', '', 'BOF') }}}
-{{{ ansible_remove_pam_module_option_configuration('/etc/pam.d/login', 'session', '', 'pam_lastlog.so', 'silent') }}}
-{{% else %}}
-{{{ ansible_ensure_pam_module_configuration('/etc/pam.d/postlogin', 'session', 'required', 'pam_lastlog.so', 'showfailed', '', 'BOF') }}}
-{{{ ansible_remove_pam_module_option_configuration('/etc/pam.d/postlogin', 'session', '', 'pam_lastlog.so', 'silent') }}}
-{{% endif %}}
+{{%- if product in ["sle12", "sle15"] or 'ubuntu' in product %}}
+{{%- set pam_lastlog_path = "/etc/pam.d/login" %}}
+{{%- set control = "optional" %}}
```
Comment (dodys) [inline linux_os/guide/system/accounts/accounts-pam/display_login_attempts/ansible/shared.yml:9]: control for ubuntu is `required`

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/bash/shared.sh [hunk 7]
```diff
@@ -1,9 +1,13 @@
...
-{{% if product in ["sle12", "sle15"] or "ubuntu" in product %}}
-{{{ bash_ensure_pam_module_configuration('/etc/pam.d/login', 'session', 'required', 'pam_lastlog.so', 'showfailed', '', 'BOF') }}}
-{{{ bash_remove_pam_module_option_configuration('/etc/pam.d/login', 'session', '', 'pam_lastlog.so', 'silent') }}}
-{{% else %}}
-{{{ bash_ensure_pam_module_configuration('/etc/pam.d/postlogin', 'session', 'required', 'pam_lastlog.so', 'showfailed', '', 'BOF') }}}
-{{{ bash_remove_pam_module_option_configuration('/etc/pam.d/postlogin', 'session', '', 'pam_lastlog.so', 'silent') }}}
-{{% endif %}}
+{{%- if product in ["sle12", "sle15"] or 'ubuntu' in product %}}
+{{%- set pam_lastlog_path = "/etc/pam.d/login" %}}
+{{%- set control = "optional" %}}
```
Comment (dodys) [inline linux_os/guide/system/accounts/accounts-pam/display_login_attempts/bash/shared.sh:5]: control for ubuntu is `required`

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/rule.yml [hunk 8]
```diff
@@ -9,20 +9,17 @@ documentation_complete: true
...
-    using <tt>pam_lastlog</tt>, add or correct the <tt>pam_lastlog</tt>
-    settings in
-    <tt>{{{ pam_lastlog_path }}}</tt> to read as follows:
-    <pre>session     required pam_lastlog.so showfailed</pre>
-    And make sure that the <tt>silent</tt> option is not set for
-    <tt>pam_lastlog</tt> module.
+    To configure the system to notify users of last logon/access using <tt>pam_lastlog</tt>,
+    add or correct the <tt>pam_lastlog</tt> settings in <tt>{{{ pam_lastlog_path }}}</tt>
+    to read as follows:
+    <pre>session     [default=1]    pam_lastlog.so nowtmp showfailed</pre>
```
Comment (dodys) [inline linux_os/guide/system/accounts/accounts-pam/display_login_attempts/rule.yml:15]: for now the Ubuntu 20.04 STIG benchmark still mentions:
`session required pam_lastlog.so showfailed`

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/commented_line.fail.sh [hunk 9]
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
-# platform = Oracle Linux 7,Red Hat Enterprise Linux 7,multi_platform_ubuntu
+# platform = multi_platform_ubuntu,Oracle Linux 7,Red Hat Enterprise Linux 7
```
Comment (dodys) [inline linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/commented_line.fail.sh:2]: missing SLE platform?

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/correct_value.pass.sh [hunk 10]
```diff
@@ -1,10 +1,17 @@
...
-echo "session required pam_lastlog.so showfailed" >> /etc/pam.d/postlogin
+{{% set pam_lastlog_path = "/etc/pam.d/postlogin" %}}
 {{% endif %}}
+
+rm -f {{{ pam_lastlog_path }}}
+
+cat <<EOF > {{{ pam_lastlog_path }}}
+session     optional                   pam_umask.so silent
+session     [success=1 default=ignore] pam_succeed_if.so service !~ gdm* service !~ su* quiet
+session     [default=1]                pam_lastlog.so nowtmp showfailed
```
Comment (dodys) [inline linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/correct_value.pass.sh:15]: perhaps make control part of the if above and refer to it here?

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/no_silent_all.pass.sh [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+{{% else %}}
+{{% set pam_lastlog_path = "/etc/pam.d/postlogin" %}}
+{{% endif %}}
+
+rm -f {{{ pam_lastlog_path }}}
+
+cat <<EOF > {{{ pam_lastlog_path }}}
+session     optional                   pam_umask.so silent
+session     [success=1 default=ignore] pam_succeed_if.so service !~ gdm* service !~ su* quiet
+session     [default=1]                pam_lastlog.so nowtmp showfailed
```
Comment (dodys) [inline linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/no_silent_all.pass.sh:15]: same as above

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/no_space_before_showfailed.fail.sh [hunk 12]
```diff
@@ -1,17 +1,17 @@
...
-    # or modify
-    sed -Ei --follow-symlinks 's/^\s*session\s+required\s+pam_lastlog\.so(\s.*)?\sshowfailed($|\s)/session	required		pam_lastlog.so\1	silent=showfailed\2/}' $CUSTOM_POSTLOGIN
-fi
-authselect apply-changes -b
+rm -f {{{ pam_lastlog_path }}}
+
+cat <<EOF > {{{ pam_lastlog_path }}}
+session     optional                   pam_umask.so silent
+session     [success=1 default=ignore] pam_succeed_if.so service !~ gdm* service !~ su* quiet
+session     [default=1]                pam_lastlog.so nowtmp=showfailed
```
Comment (dodys) [inline linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/no_space_before_showfailed.fail.sh:15]: same as above

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/no_space_before_silent.pass.sh [hunk 13]
```diff
@@ -1,20 +1,17 @@
...
-    # or modify
-    sed -Ei --follow-symlinks 's/^\s*session\s+required\s+pam_lastlog\.so(\s.*)?\ssilent($|\s)/session	required		pam_lastlog.so\1	unknown=silent\2/}' $CUSTOM_POSTLOGIN
-fi
-authselect apply-changes -b
+rm -f {{{ pam_lastlog_path }}}
+
+cat <<EOF > {{{ pam_lastlog_path }}}
+session     optional                   pam_umask.so silent
+session     [success=1 default=ignore] pam_succeed_if.so service !~ gdm* service !~ su* quiet
+session     [default=1]                pam_lastlog.so unknown=silent nowtmp showfailed
```
Comment (dodys) [inline linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/no_space_before_silent.pass.sh:15]: same as above

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/silent_present.fail.sh [hunk 14]
```diff
@@ -9,10 +9,9 @@
...
-# so let's add more lines there
-echo "session     required                   pam_lastlog.so showfailed" >> {{{ pam_lastlog_path }}}
-echo "session     optional                   pam_umask.so silent" >> {{{ pam_lastlog_path }}}
-echo "session     [success=1 default=ignore] pam_succeed_if.so service !~ gdm* service !~ su* quiet" >> {{{ pam_lastlog_path }}}
-echo "session     [default=1]                pam_lastlog.so nowtmp showfailed" >> {{{ pam_lastlog_path }}}
-echo "session     optional                   pam_lastlog.so silent noupdate showfailed" >> {{{ pam_lastlog_path }}}
+cat <<EOF > {{{ pam_lastlog_path }}}
+session     optional                   pam_umask.so silent
+session     [success=1 default=ignore] pam_succeed_if.so service !~ gdm* service !~ su* quiet
+session     [default=1]                pam_lastlog.so nowtmp silent showfailed
```
Comment (dodys) [inline linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/silent_present.fail.sh:15]: same as above

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/silent_present.fail.sh [hunk 15]
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
-# platform = Oracle Linux 7,Red Hat Enterprise Linux 7,multi_platform_ubuntu
+# platform = multi_platform_ubuntu,Oracle Linux 7,Red Hat Enterprise Linux 7
```
Comment (dodys) [inline linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/silent_present.fail.sh:2]: missing SLE platform?

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/wrong_value.fail.sh [hunk 16]
```diff
@@ -1,9 +1,17 @@
 #!/bin/bash
-# platform = Oracle Linux 7,Red Hat Enterprise Linux 7
+# platform = multi_platform_ubuntu,Oracle Linux 7,Red Hat Enterprise Linux 7
```
Comment (dodys) [inline linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/wrong_value.fail.sh:2]: missing SLE platform?

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/wrong_value.fail.sh [hunk 17]
```diff
@@ -1,9 +1,17 @@
...
+{{% else %}}
+{{% set pam_lastlog_path = "/etc/pam.d/postlogin" %}}
+{{% endif %}}
+
+rm -f {{{ pam_lastlog_path }}}
+
+cat <<EOF > {{{ pam_lastlog_path }}}
+session     optional                   pam_umask.so silent
+session     [success=1 default=ignore] pam_succeed_if.so service !~ gdm* service !~ su* quiet
+session     [default=1]                pam_lastlog.so wrong_value
```
Comment (dodys) [inline linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/wrong_value.fail.sh:15]: same as previously mentioned

**Review by dodys — CHANGES_REQUESTED** [review]
some small details regarding ubuntu specifically

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/commented_line.fail.sh [hunk 18]
```diff
@@ -10,12 +10,8 @@
...
-#%PAM-1.0
-# This file is auto-generated.
-# User changes will be destroyed the next time authconfig is run.
-
-
+session     optional                   pam_umask.so silent
 session     [success=1 default=ignore] pam_succeed_if.so service !~ gdm* service !~ su* quiet
-session     [default=1]   pam_lastlog.so nowtmp showfailed
-session     optional      pam_lastlog.so silent noupdate showfailed
+#session     [default=1]                pam_lastlog.so nowtmp showfailed
```
[the remark this answers is not in the document]
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/commented_line.fail.sh:15]: Good point. I will update.

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/oval/shared.xml [hunk 19]
```diff
@@ -4,31 +4,29 @@
...
-  </ind:textfilecontent54_object>
-
-  <!-- DISA STIG forbids the 'silent' option for SLE12/SLE15, OL7, RHEL7/RHEL8 -->
-  <ind:textfilecontent54_test check="all" check_existence="none_exist" comment="Forbid 'silent' option for pam_lastlog" id="test_display_login_attempts_silent" version="1">
-    <ind:object object_ref="obj_display_login_attempts_silent" />
-  </ind:textfilecontent54_test>
-  <ind:textfilecontent54_object id="obj_display_login_attempts_silent" version="1">
-    <ind:filepath>{{{ pam_lastlog_path }}}</ind:filepath>
-    <ind:pattern operation="pattern match">^\s*session\s+.*\s+pam_lastlog\.so(?:\s+[\w=]+)*\s+silent(\s|$)</ind:pattern>
+    <ind:pattern operation="pattern match">^\s*session\s+.*\s+pam_lastlog\.so\b(?!.*\ssilent\s).*\sshowfailed\s.*$</ind:pattern>
```
[the remark this answers is not in the document]
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/oval/shared.xml:29]: I don't think so. It is intentional to not check the control here as many different combinations will work without technical impact and it would be very complex to satisfy each possible combination. The control is more relevant for remediation, in rare cases the default lines are not showing the last logon information.

[the remark this answers is not in the document]
Comment (dodys) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/oval/shared.xml:29]: agreed

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/rule.yml (hunk 8 above)
Comment (freddieRv) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/rule.yml:15]: The STIGs for OL also have the same config entry:
```
session required pam_lastlog.so showfailed
```

Reading the [man pages](https://www.man7.org/linux/man-pages/man5/pam.conf.5.html) there is an equivalent control using the `[...]` syntax:
```
Each of the four keywords: required; requisite; sufficient; and
optional, have an equivalent expression in terms of the [...]
syntax. They are as follows:

       required
           [success=ok new_authtok_reqd=ok ignore=ignore default=bad]

       requisite
           [success=ok new_authtok_reqd=ok ignore=ignore default=die]

       sufficient
           [success=done new_authtok_reqd=done default=ignore]

       optional
           [success=ok new_authtok_reqd=ok default=ignore]
```

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/rule.yml (hunk 2 above)
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/rule.yml]: Fixed

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/rule.yml (hunk 3 above)
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/rule.yml]: Fixed

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/correct_value.pass.sh (hunk 4 above)
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/correct_value.pass.sh]: Fixed

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/default_config.pass.sh (hunk 5 above)
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/default_config.pass.sh]: Fixed

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/ansible/shared.yml (hunk 6 above)
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/ansible/shared.yml]: Fixed

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/bash/shared.sh (hunk 7 above)
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/bash/shared.sh]: Fixed

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/commented_line.fail.sh (hunk 9 above)
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/commented_line.fail.sh]: Fixed

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/silent_present.fail.sh (hunk 15 above)
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/silent_present.fail.sh]: Fixed

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/wrong_value.fail.sh (hunk 16 above)
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/wrong_value.fail.sh]: Fixed

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/no_space_before_showfailed.fail.sh [hunk 20]
```diff
@@ -1,17 +1,17 @@
 #!/bin/bash
-# packages = authselect
-# platform = Oracle Linux 8,Oracle Linux 9,Red Hat Enterprise Linux 8,Red Hat Enterprise Linux 9,multi_platform_fedora
+# platform = multi_platform_fedora,multi_platform_ubuntu,Oracle Linux 7,Red Hat Enterprise Linux 7
```
[the remark this answers is not in the document]
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/no_space_before_showfailed.fail.sh]: Fixed

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/correct_value.pass.sh (hunk 10 above)
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/correct_value.pass.sh:15]: Instead of creating more conditionals, I ensured two `pass` scripts with different controls. It should be enough for testing purposes. https://github.com/ComplianceAsCode/content/pull/10979/commits/21ae1e154f60c185a96daf401782665115e15242

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/no_space_before_showfailed.fail.sh (hunk 12 above)
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/no_space_before_showfailed.fail.sh:15]: https://github.com/ComplianceAsCode/content/pull/10979#discussion_r1308395060

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/no_space_before_silent.pass.sh (hunk 13 above)
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/no_space_before_silent.pass.sh:15]: https://github.com/ComplianceAsCode/content/pull/10979#discussion_r1308395060

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/silent_present.fail.sh (hunk 14 above)
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/silent_present.fail.sh:15]: https://github.com/ComplianceAsCode/content/pull/10979#discussion_r1308395060

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/wrong_value.fail.sh (hunk 17 above)
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/wrong_value.fail.sh:15]: https://github.com/ComplianceAsCode/content/pull/10979#discussion_r1308395060

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/commented_line.fail.sh (hunk 18 above)
[the remark this answers is not in the document]
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/tests/commented_line.fail.sh:15]: https://github.com/ComplianceAsCode/content/pull/10979#discussion_r1308395060

File: linux_os/guide/system/accounts/accounts-pam/display_login_attempts/rule.yml (hunk 8 above)
Comment (marcusburghardt) [reply linux_os/guide/system/accounts/accounts-pam/display_login_attempts/rule.yml]: @freddieRv , we use these two syntaxes in some PAM rules, but in the context of `pam_lastlog` the control is not so relevant to achieve the expected outcome. Many combinations would work fine. In any case, in my last commits I updated the verbiage in rule descriptions to be more flexible. This is similar to my comment in https://github.com/ComplianceAsCode/content/pull/10979#discussion_r1307627997.

**Review by dodys — APPROVED** [review]
LGTM, thanks for addressing the comments!

**Review by freddieRv — APPROVED** [review]
Changes LGTM. Thanks for addressing my comment @marcusburghardt

**Comment by Mab879** [comment]
Thanks all for your reviews!
